packages/@aws-cdk/aws-sns/lib/subscription.ts

```diff
@@ -103,39 +110,69 @@ export class Subscription extends Resource {
       if (Object.keys(props.filterPolicy).length > 5) {
         throw new Error('A filter policy can have a maximum of 5 attribute names.');
       }
-
-      this.filterPolicy = Object.entries(props.filterPolicy)
-        .reduce(
-          (acc, [k, v]) => ({ ...acc, [k]: v.conditions }),
-          {},
-        );
-
-      let total = 1;
-      Object.values(this.filterPolicy).forEach(filter => { total *= filter.length; });
-      if (total > 100) {
-        throw new Error(`The total combination of values (${total}) must not exceed 100.`);
-      }
+      this.filterPolicy = this.buildFilterPolicy(props.filterPolicy);
     }
 
     if (props.protocol === SubscriptionProtocol.FIREHOSE && !props.subscriptionRoleArn) {
       throw new Error('Subscription role arn is required field for subscriptions with a firehose protocol.');
     }
 
     this.deadLetterQueue = this.buildDeadLetterQueue(props);
-
     new CfnSubscription(this, 'Resource', {
       endpoint: props.endpoint,
       protocol: props.protocol,
       topicArn: props.topic.topicArn,
       rawMessageDelivery: props.rawMessageDelivery,
       filterPolicy: this.filterPolicy,
+      filterPolicyScope: props.filterPolicyScope,
       region: props.region,
       redrivePolicy: this.buildDeadLetterConfig(this.deadLetterQueue),
       subscriptionRoleArn: props.subscriptionRoleArn,
     });
 
   }
 
+  // FIRST OPTION DFS
+  private buildFilterPolicy(filterPolicy: any, depth = 1, totalCombinationValues = [1]) {
+    Object.entries(filterPolicy).forEach(([key, value]) => {
+      if (value instanceof SubscriptionFilter) {
+        filterPolicy[key] = value.conditions;
+        totalCombinationValues[0] *= value.conditions.length * depth;
+      } else {
+        this.buildFilterPolicy(value, depth + 1, totalCombinationValues);
+      }
+    });
+    if (totalCombinationValues[0] > 100) {
+      throw new Error(`The total combination of values (${totalCombinationValues}) must not exceed 100.`);
+    }
+    return filterPolicy;
+  };
+
+  // SECOND OPTION BFS
+  // private buildFilterPolicy(policyFilter: SubscriptionFilterPolicy) {
+  //   const queue: any = [policyFilter];
+  //   let totalCombinationValues = 1;
+  //   let depth = 1;
+  //   while (queue.length > 0) {
+  //     const current = queue.shift();
+  //     let levelSize = queue.length;
+  //     while (levelSize > 0) {
+
+  //       if (val instanceof SubscriptionFilter) {
+  //         totalCombinationValues *= val.conditions.length * depth;
+  //         current[key] = val.conditions;
+  //       } else {
+  //         queue.push(val);
+  //       }
+  //     }
+  //     depth++;
+  //   }
+  //   if (totalCombinationValues > 100) {
+  //     throw new Error(`The total combination of values (${totalCombinationValues}) must not exceed 100.`);
+  //   }
+  //   return policyFilter;
+  // }
+
```

Comment:
Prefer recursive DFS for its simplicity. Although BFS is non recursive, it does get quite complicated with calculating the depth / maintaining queue of nodes.

Reply:
Calculations for policy constraints is as followed here: https://docs.aws.amazon.com/sns/latest/dg/subscription-filter-policy-constraints.html.